Modelica/Blocks/Logical.mo

```diff
@@ -973,6 +973,41 @@ parameter \"terminationText\".
 
 </html>"));
   end TerminateSimulation;
+
+  block AssertCondition "Trigger error (or warning) and print message if assertion condition is not fulfilled"
+
+    Modelica.Blocks.Interfaces.BooleanOutput condition=true
```

Comment:
Can anyone explain why this (and the same in `TerminateSimulation`) is an output and not an input? I mean, what's the point of having an output that can only have the value is `true` (or stays `false` in the case of `TerminateSimulation`). Sure, with `AssertionLevel.warning` it can change back and forth, but it would seem more useful to me if the blocks could be controlled via an input connector.

I suppose one could also arrange for a conditional input connector, allowing the same block to be controlled by either an expression or an input?

Reply:
While it may not fully answer the question, I'll give you my comments.

I expect I would rarely, if ever, use the external output connection as is. Maybe there are some unique use cases possible when using the warning assertion level as you hinted. With that said, I added it for two reasons: a) to be in alignment with the TerminateSimulation block, and b) because why not, it doesn't hurt and it may have some use cases that are helpful that I'm not aware of.

Alternatively, if the block had an input connector, I would say 99% of the time I would always just have a "BooleanExpression" source block immediately upstream, in which case the design as-is is preferred for simplicity and ease of use.

Another comment is that, to me, the normal usage of this block should feel comparable to writing an assert statement in the equation section, but the main difference is that I've done it in the graphical layer rather than in the equation section.

Reply:
Sure, I can also imagine that the more common use case would be to write an expression, and the question of how to control the block reminds me of some recent mention of the addition variant of a `Gain` block: we currently have `Add`, `Procuct`, and `Gain`, but no `Shift` (or `Offset`, etc) block. Just as you say regarding the use of `AssertCondition`, it is quite inconvenient to always have to use the combination of `Constant` and `Add` in order to shift a signal.  The situation is not identical, as the `Gain` has a parameter that allows the factor to be modified at initialization, but I still think it illustrates the desire to have alternatives for both the input and declaration equation variants.

Hence, I agree that only supporting the use of an input connector would be poor design. This is what I have in mind:
```
block AssertCondition "Trigger error (or warning) and print message if assertion condition is not fulfilled"
  parameter Boolean enableInput = false "Enable use of input connector instead of condition expression";
  Modelica.Blocks.Interfaces.BooleanOutput condition = true if not enableInput "Trigger error (or warning) and print message when condition becomes false" annotation(Dialog(enable = not enableInput), Placement(transformation(extent = {{200, -10}, {220, 10}})));
  Modelica.Blocks.Interfaces.BooleanInput u if enableInput "Trigger error (or warning) and print message when input becomes false" annotation(Dialog(enable = enableInput), Placement(transformation(extent = {{200, -10}, {220, 10}}, origin = {0, 0}, rotation = 0), visible = true, iconTransformation(origin = {-200, 0}, extent = {{-20, -10}, {0, 10}}, rotation = 0)));
  parameter String message = "... assertion condition no longer true" "Text that will be displayed when assertion condition is false";
  parameter AssertionLevel assertionLevel = AssertionLevel.error "Select warning (1) or error (2)";
protected
  Modelica.Blocks.Interfaces.BooleanInput cond;
equation
  connect(condition, cond);
  connect(u, cond);
  assert(cond, message, assertionLevel);
  annotation(Icon(coordinateSystem(preserveAspectRatio = true, extent = {{-200, -20}, {200, 20}}, initialScale = 0.2, grid = {10, 10}), graphics = {Rectangle(visible = true, fillColor = {235, 235, 235}, fillPattern = FillPattern.Solid, lineThickness = 5, borderPattern = BorderPattern.Raised, extent = {{-200, -20}, {200, 20}}), Text(visible = not enableInput, origin = {-15.556, 0}, extent = {{-184.444, -15}, {215.556, 15}}, textString = "assert(%condition)"), Text(visible = true, textColor = {0, 0, 255}, extent = {{-200, 22}, {200, 46}}, textString = "%name"), Text(visible = enableInput, origin = {-15.556, 0}, extent = {{-184.444, -15}, {215.556, 15}}, textString = "assert")}), Documentation(info = "<html>
<p>In the parameter menu, a <strong>time varying</strong> expression can be defined via variable <strong>condition</strong>, for example &quot;condition = x &lt; 0&quot;, where &quot;x&quot; is a variable that is declared in the model in which the &quot;AssertCondition&quot; block is present. If this expression becomes <strong>false</strong>, the simulation will either abort with a message indicating the cause of the error (if assertion level is set to error) or not abort with a message indicating the cause of the warning (if assertion level is set to warning).</p>
</html>"));
end AssertCondition;
```

Note that I have also modified the icon to make it more distinguishable from the `TerminateSimulation`, and the contents of the icon are controlled by the `enableInput` parameter.

Reply:
To me both this and `TerminateSolution` somewhat break the separation of concerns.

The general idea is to have connections for propagating signals, and we have `Modelica.Blocks.Sources.BooleanExpression` (etc) if we want to write an expression and generate a signal.

Having (optional) expression-code in many blocks lead to less clear solutions. I will comment on one of them.

Reply:
Thinking more based on #3986 I believe that there is a clear graphical advantage of using BooleanExpression to generate the signal and then having an assert-block (and terminate-block); even if it at first seems somewhat odd.

The advantage is that we can have a symbol for assert (and terminate) with fix aspect ratio, and a pure expression-block that can be adapted to the size of the expression.

Reply:
@henrikt-ma While I understand your suggestion, I personally think it overcomplicates what should be a simple block and the block visual/icon would seem a little out of place when using it with the input connector enabled. With that said, I think I'm coming around to the idea of just having an input connector always enabled and without the time-varying dialog input, as suggested by @HansOlsson. What are your thoughts?

@HansOlsson While I think it adds at least a *little* overhead to have to use two blocks for the typical use case (BooleanExpression + AssertCondition), I am coming around to the idea that maybe that approach is a bit more elegant overall.

Assuming there is an agreement, what is the best way to proceed? It seems to me that this block and `TerminateSimulation` should use the same overall approach. If agreed, it seems one of two options could be taken:
Option 1:
1) Complete this PR mostly as is (i.e. `AssertCondition` similar to `TerminateSimulation`)
2) Later, issue another PR that changes both `AssertCondition` and `TerminateSimulation` to the approach that uses only an input connector

Option 2:
1) Rework this PR to revise `AssertCondition` so that it uses only an input connector
2) Later, issue another PR that brings `TerminateSimulation` into alignment with `AssertCondition`

...this is all assuming there is agreement. Any other thoughts/suggestions?

Reply:
I would believe 2 is best. The reason is that when changing TerminateSimulation we need to consider changing all uses - and to avoid breaking compatibility we might even keep both; it's good if we don't have to do that for AssertCondition as well.

Reply:
I don't have strong opinions on this matter.  I just note that I often find it cumbersome to use a `Constant` block when doing things that would have been solved compactly using the `Gain` block in case of multiplication, and that there would be a way to make one block covering both usage patterns.

Anyway, to always control the block via an input at least resolves the confusion about why the controlling expression is attached to an output connector.

Reply:
Block has been restructured to use input signal rather than dialog input